### PR TITLE
Ensure PVs and PVCs remain bound when doing a restore

### LINKED ISSUE #2570: Velero randomly restoring another namespaces PVCs on a backup.
**What steps did you take and what happened:**
We schedule a backup when creating our development environment. After the environment is created, velero somehow replaces the new PVC with a cloned PVC instead of keeping the new PVC.

This is a huge issue for us since we do not want PVCs randomly being attached. I am unsure why it's doing it, we are not issuing any restore commands. The only command I see that we are issuing is a backup schedule.

At first I thought maybe our ingress was just serving the wrong environment, but when I run `kubetl get pvc` on the new environment, I see velero-clone in my PVC names:

kubectl get pvc -n wp-bf083a8f-1ac7-416c-bac5-234f586c4440
```
NAME            STATUS   VOLUME                                              CAPACITY   ACCESS MODES   STORAGECLASS   AGE
mysql-mysql-0   Bound    velero-clone-1a2a1a6c-915f-4096-a693-cff87b547190   5Gi        RWO            ssd            111m
site-code       Bound    pvc-b758f25c-9e94-11ea-9504-42010aa20135            5Gi        RWO            ssd            111m
```

I wanted to check the restore logs, but `velero restore get` is returning nothing for some reason. Is there another way to get a list of restores and their logs? Maybe I'm running into a bug?

**What did you expect to happen:**
I did not expect a backup schedule to trigger a PVC clone. I expected the backup schedule to just create a backup as usual.

**The output of the following commands will help us better understand what's going on**:
(Pasting long output into a [GitHub gist](https://gist.github.com) or other pastebin is fine.)

* `kubectl logs deployment/velero -n velero`
https://pastebin.com/Z1MgNT20

* `velero backup describe <backupname>` or `kubectl get backup/<backupname> -n velero -o yaml`

```
Name:         wp-bf083a8f-1ac7-416c-bac5-234f586c4440-20200525143337
Namespace:    velero
Labels:       velero.io/schedule-name=wp-bf083a8f-1ac7-416c-bac5-234f586c4440
              velero.io/storage-location=default
Annotations:  <none>

Phase:  Completed

Namespaces:
  Included:  wp-bf083a8f-1ac7-416c-bac5-234f586c4440
  Excluded:  <none>

Resources:
  Included:        *
  Excluded:        <none>
  Cluster-scoped:  auto

Label selector:  <none>

Storage Location:  default

Snapshot PVs:  true

TTL:  336h0m0s

Hooks:  <none>

Backup Format Version:  1

Started:    2020-05-25 10:33:37 -0400 EDT
Completed:  2020-05-25 10:33:58 -0400 EDT

Expiration:  2020-06-08 10:33:37 -0400 EDT

Persistent Volumes:  2 of 2 snapshots completed successfully (specify --details for more information)
```

* `velero backup logs <backupname>`

https://pastebin.com/sVuiWjGk

* `velero restore describe <restorename>` or `kubectl get restore/<restorename> -n velero -o yaml`

Cannot describe restore. Velero restore get is not showing any entries.

* `velero restore logs <restorename>`

Cannot get restore logs. Velero restore get is not showing any entries.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Velero version (use `velero version`): 
```
Client:
	Version: v1.3.1
	Git commit: -
Server:
	Version: v1.3.2
```
- Velero features (use `velero client config get features`): 

```
features: <NOT SET>
```

- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.2", GitCommit:"c97fe5036ef3df2967d086711e6c0c405941e14b", GitTreeState:"clean", BuildDate:"2019-10-15T23:41:55Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.10-gke.36", GitCommit:"34a615f32e9a0c9e97cdb9f749adb392758349a6", GitTreeState:"clean", BuildDate:"2020-04-06T16:33:17Z", GoVersion:"go1.12.12b4", Compiler:"gc", Platform:"linux/amd64"}
```
- Kubernetes installer & version:
GKE (Google Cloud)

- Cloud provider or hardware configuration:
Google Cloud
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by nrb**
Fixes #2570 

This PR makes the following changes:

- Clears out the `Spec.ClaimRef.UID` and `Spec.ClaimRef.ResourceVersion` fields on a PersistentVolume when restoring, rather than completely removing the ClaimRef. This leaves the `Namespace` and `Name` intact, allowing the PersistentVolume to be re-bound to the newly restored PersistentVolumeClaim. Not doing so may lead to issues with data leakage as it's possible Kubernetes could bind the PersistentVolume to a PersistententVolumeClaim in another namespace if it matches the right conditions. See issue #2570 for discussion of this as well as [this upstream Kubernetes comment](https://github.com/kubernetes/kubernetes/issues/48609#issuecomment-314066616)
- Clears out the binding annotations set by the Kubernetes controller on PersistentVolumes and PersistentVolumeClaims. This is in order ensure specific logic is triggered that re-binds the objects based on name alone, rather than UID. Code for the controller in question can be found in [this file](https://github.com/kubernetes/kubernetes/blob/b86e725694e81b48ea36b11f647f28cf70a66210/pkg/controller/volume/persistentvolume/pv_controller.go)
- Updates most tests with the new logic.

The main cases of restoring a snapshotted volume against a cluster and remapping a volume have been tested, but reprovisioning volumes with restic has not yet been tested.

The issues that this fixes are not always encountered, but can happen on a busy cluster.

### CODE REVIEW

**Comment by nrb** [comment]
cc @arianitu

**Comment by nrb** [comment]
Based on my local `make test` runs, CI should be fixed, but I'll still need to check against restic on a live cluster.

File: pkg/restore/restore.go [hunk 1]
```diff
@@ -949,6 +959,7 @@ func (ctx *restoreContext) restoreItem(obj *unstructured.Unstructured, groupReso
 			return warnings, errs
 
 		case hasDeleteReclaimPolicy(obj.Object):
+			fmt.Printf("Has a delete policy!")
```
Comment (arianitu) [inline pkg/restore/restore.go:962]: probably want to remove this

File: pkg/restore/restore.go [hunk 2]
```diff
@@ -1215,6 +1233,43 @@ func shouldRenamePV(ctx *restoreContext, obj *unstructured.Unstructured, client
...
+
+	targetNS, ok := ctx.restore.Spec.NamespaceMapping[pv.Spec.ClaimRef.Namespace]
+
+	if !ok {
+		ctx.log.Debugf("Persistent volume does not need to have the claimRef.namespace remapped because it's not claimed by a PVC in a namespace that's being remapped")
+		return false, nil
+	}
+
+	var err error
+	if ok {
```
Comment (arianitu) [inline pkg/restore/restore.go:1263]: ok is guaranteed here, should we check again or just call `err = unstructured.SetNestedField(obj.Object, targetNS, "spec", "claimRef", "namespace") ` directly unless this is a coding practice on velero?

**Review by arianitu — COMMENTED** [review]
I do not know the codebase very well, so I did a basic sanity check, will need someone who knows the implications a bit better than I do, but looks good to me.

File: pkg/restore/restore.go (hunk 1 above)
Comment (nrb) [reply pkg/restore/restore.go:962]: Though I had gotten all my debug prints, thanks :)

**Comment by nrb** [comment]
@arianitu Thanks for taking a look, I appreciate it! I will incorporate this feedback.

@vmware-tanzu/velero-maintainers I've been able to confirm restic support works as expected - new, empty PVs are provisioned and restic will restore data into them.

**Review by carlisia — APPROVED** [review]
This lgtm 👍

**Review by ashish-amarnath — APPROVED** [review]
LGTM!

**Comment by arianitu** [comment]
Thank you guys so much, very appreciated! @nrb @carlisia @ashish-amarnath, I owe you a beer!

**Pull request merged**